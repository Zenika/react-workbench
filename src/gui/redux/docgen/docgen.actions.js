export const REQUEST_DOCGEN = 'REQUEST_DOCGEN'
export const RECEIVED_DOCGEN = 'RECEIVED_DOCGEN'
export const RECEIVED_EMPTY_DOCGEN = 'RECEIVED_EMPTY_DOCGEN'
export const FAILED_DOCGEN = 'FAILED_DOCGEN'

export const fetchDocgen = () => async (dispatch) => {
  dispatch({ type: REQUEST_DOCGEN })
  try {
    const response = await fetch('/api/docgen')
    const json = await response.json()
    if (json && json.length > 0) {
      dispatch({ type: RECEIVED_DOCGEN, payload: json[0] })
    } else {
      dispatch({ type: RECEIVED_EMPTY_DOCGEN })
    }
  } catch (error) {
    dispatch({ type: FAILED_DOCGEN, error })
  }
}
