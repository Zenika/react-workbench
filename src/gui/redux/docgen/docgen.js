import factory from 'trampss-redux-factory'

export default factory({ name: 'docgen', type: 'uniq' })
