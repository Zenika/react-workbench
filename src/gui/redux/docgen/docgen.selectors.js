import { createSelector } from 'reselect'
import { docgen } from '../reducers'

const getProps = state => docgen.get()(state).props || {}
export const getPropsKeys = createSelector(getProps, p => Object.keys(p))
export const getProp = name => createSelector(getProps, p => p[name])
