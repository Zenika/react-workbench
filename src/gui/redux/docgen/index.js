export { default } from './docgen'
export * from './docgen.selectors'
export * from './docgen.api'
