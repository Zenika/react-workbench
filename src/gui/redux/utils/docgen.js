const convertDocgenValue = (value, type) => {
  switch (type) {
    case 'bool':
      return value === 'true'
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'symbol':
      return value
    default:
      return eval(value) // eslint-disable-line no-eval
  }
}

const docgenToModel = (docgen) => {
  return Object.keys(docgen.props || {}).reduce((obj, key) => {
    const prop = docgen.props[key]
    const type = prop.type && prop.type.name
    const value = prop.defaultValue && prop.defaultValue.value
    return { ...obj, [key]: convertDocgenValue(value, type) }
  }, {})
}

const convertToGuiValue = (value, type) => {
  switch (type) {
    case 'array':
    case 'object':
      return JSON.stringify(value)
    case 'func':
      return eval(value) // eslint-disable-line no-eval
    default:
      return value
  }
}

const convertFromGuiValue = (value, type) => {
  switch (type) {
    case 'bool':
      return !!value
    case 'object':
      return eval(`Object(${value})`) // eslint-disable-line no-eval
    case 'array':
    case 'func':
      return eval(value) // eslint-disable-line no-eval
    default:
      return value
  }
}

export { convertDocgenValue, docgenToModel, convertToGuiValue, convertFromGuiValue }
