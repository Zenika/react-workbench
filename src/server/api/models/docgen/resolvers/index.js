const componentResolver = require('./componentResolver')
const indexResolver = require('./indexResolver')

module.exports = Object.assign(
  {},
  { componentResolver },
  { indexResolver }
)
