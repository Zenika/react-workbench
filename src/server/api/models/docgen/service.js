const fs = require('fs')
const p = require('path')
const docgen = require('react-docgen')

const utils = require('react-docgen/dist/utils')
const resolveHOC = require('react-docgen/dist/utils/resolveHOC.js').default

const isComponentDefinition = (path) => {
  return utils.isReactCreateClassCall(path)
    || utils.isReactComponentClass(path)
    || utils.isStatelessComponent(path)
}

const resolveDefinition = (definition, types) => {
  if (utils.isReactCreateClassCall(definition)) {
    const resolvedPath = utils.resolveToValue(definition.get('arguments', 0))
    if (types.ObjectExpression.check(resolvedPath.node)) {
      return resolvedPath
    }
  } else if (utils.isReactComponentClass(definition)) {
    utils.normalizeClassDefinition(definition)
    return definition
  } else if (utils.isStatelessComponent(definition)) {
    return definition
  }
  return null
}

const componentResolver = (ast, recast) => {
  const types = recast.types.namedTypes
  const components = []

  const exportDeclaration = (path) => {
    const definitions = utils.resolveExportDeclaration(path, types)
      .reduce((acc, definition) => {
        if (isComponentDefinition(definition)) {
          acc.push(definition)
        } else {
          const resolved = utils.resolveToValue(resolveHOC(definition))
          if (resolved.node.type === 'ImportDeclaration') {
            // Manage imported components in HOC from different file
            const importType = { path: resolved.node.source.value }
            throw importType
          }

          if (isComponentDefinition(resolved)) {
            acc.push(resolved)
          }
        }
        return acc
      }, [])
      .map(definition => resolveDefinition(definition, types))

    if (definitions.length === 0) {
      return false
    }
    definitions.forEach((definition) => {
      if (definition && components.indexOf(definition) === -1) {
        components.push(definition)
      }
    })
    return false
  }

  const assignmentExpression = (path) => {
    let resolvedPath = path
    // Ignore anything that is not `exports.X = ...;` or
    // `module.exports = ...;`
    if (!utils.isExportsOrModuleAssignment(resolvedPath)) {
      return false
    }
    // Resolve the value of the right hand side. It should resolve to a call
    // expression, something like React.createClass
    resolvedPath = utils.resolveToValue(resolvedPath.get('right'))
    if (!isComponentDefinition(resolvedPath)) {
      resolvedPath = utils.resolveToValue(resolveHOC(resolvedPath))
      if (!isComponentDefinition(resolvedPath)) {
        return false
      }
    }
    const definition = resolveDefinition(resolvedPath, types)
    if (definition && components.indexOf(definition) === -1) {
      components.push(definition)
    }
    return false
  }

  recast.visit(ast, {
    visitExportDeclaration: exportDeclaration,
    visitExportNamedDeclaration: exportDeclaration,
    visitExportDefaultDeclaration: exportDeclaration,
    visitAssignmentExpression: assignmentExpression,
  })

  return components
}

// FIXME not all cases of default exported elements managed
const indexResolver = (ast, recast) => {
  const exportDefaultIndex = (path) => {
    let resolvedPath = path
    // manage : 'export default XXX' (and with hoc)
    if (path.value.declaration) {
      resolvedPath = utils.resolveToValue(resolveHOC(path.get('declaration')))
      resolvedPath = utils.resolveToValue(resolvedPath.parentPath)
    }
    // by default manage : export { default } from './XXX'
    const importType = {
      path: resolvedPath.value.source.value,
    }
    throw importType
  }
  recast.visit(ast, {
    visitExportDefaultDeclaration: exportDefaultIndex,
  })
}

const resolvePath = async (path) => {
  const extensions = ['.js', '.jsx', '']
  for (let i = 0; i < extensions.length; i += 1) {
    const newpath = path + extensions[i]
    try {
      const stats = await fs.statAsync(newpath)
      if (stats.isDirectory()) {
        return resolvePath(p.resolve(path, 'index.js'))
      }
      return newpath
    } catch (e) {
      // file path not found
    }
  }
  return null
}

const resolve = async (componentPath) => {
  const resolvedPath = await resolvePath(componentPath)
  const isIndex = resolvedPath.endsWith('/index.js')
  const file = await fs.readFileAsync(resolvedPath, 'utf-8')

  try {
    return docgen.parse(file, isIndex ? indexResolver : componentResolver)
  } catch (e) {
    if (e.path) {
      const newPath = await resolvePath(p.resolve(componentPath, isIndex ? '.' : '..', e.path))
      if (newPath) {
        return resolve(newPath)
      }
    }
  }
  return null
}

module.exports = {
  resolve,
}
