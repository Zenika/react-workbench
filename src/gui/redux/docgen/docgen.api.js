import docgen from './docgen'

export const fetchDocgen = () => async (dispatch) => {
  dispatch(docgen.reset())
  try {
    const response = await fetch('/api/doc')
    const json = await response.json()
    if (json && json.length > 0) {
      dispatch(docgen.set(json[0]))
    } else {
      dispatch(docgen.set({}))
    }
  } catch (error) {
    dispatch(docgen.set({}))
  }
}
