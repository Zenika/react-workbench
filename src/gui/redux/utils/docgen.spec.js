/* eslint-env jest */
import { convertDocgenValue, docgenToModel, convertToGuiValue, convertFromGuiValue } from './docgen'

describe('gui/redux/utils', () => {
  describe('convert docgen to model structure', () => {
    it('should convert docgen to model', () => {
      const docgen = {
        props: {
          name: {
            type: { name: 'string' },
            defaultValue: { value: "'World'" },
          },
        },
      }

      expect(docgenToModel(docgen)).toMatchSnapshot()
    })
  })
  describe('convert docgen prop value to model value', () => {
    it('should convert string value', () =>
      expect(convertDocgenValue("'foo'", 'string')).toMatchSnapshot())

    it('should convert number value', () =>
      expect(convertDocgenValue('0', 'number')).toMatchSnapshot())

    it('should convert enum value', () =>
      expect(convertDocgenValue("'foo'", 'enum')).toMatchSnapshot())

    it('should convert boolean value', () =>
      expect(convertDocgenValue('false', 'bool')).toMatchSnapshot())

    it('should convert object value', () =>
      expect(convertDocgenValue("{ foo: 'bar' }", 'object')).toMatchSnapshot())

    it('should convert array value', () =>
      expect(convertDocgenValue('[1, 2]', 'array')).toMatchSnapshot())

    it('should convert symbol value', () =>
      expect(convertDocgenValue("Symbol('foo')", 'symbol')).toMatchSnapshot())

    it('should convert func value', () =>
      expect(convertDocgenValue("() => alert('test')", 'func')).toMatchSnapshot())
  })
  describe('convert model value to GUI value', () => {
    it('should convert string value', () =>
      expect(convertToGuiValue('foo', 'string')).toMatchSnapshot())

    it('should convert number value', () =>
      expect(convertToGuiValue(0, 'number')).toMatchSnapshot())

    it('should convert enum value', () =>
      expect(convertToGuiValue('foo', 'enum')).toMatchSnapshot())

    it('should convert boolean value', () =>
      expect(convertToGuiValue(false, 'bool')).toMatchSnapshot())

    it('should convert object value', () =>
      expect(convertToGuiValue({ foo: 'bar' }, 'object')).toMatchSnapshot())

    it('should convert array value', () =>
      expect(convertToGuiValue([1, 2], 'array')).toMatchSnapshot())

    it('should convert symbol value', () =>
      expect(convertToGuiValue(Symbol('foo'), 'symbol')).toMatchSnapshot())

    it('should convert func value', () =>
      expect(convertToGuiValue(() => ({}), 'func')).toMatchSnapshot())
  })
  describe('convert GUI value to model value', () => {
    it('should convert string value', () =>
      expect(convertFromGuiValue('foo', 'string')).toMatchSnapshot())

    it('should convert number value', () =>
      expect(convertFromGuiValue('0', 'number')).toMatchSnapshot())

    it('should convert enum value', () =>
      expect(convertFromGuiValue('foo', 'enum')).toMatchSnapshot())

    it('should convert boolean value', () =>
      expect(convertFromGuiValue('false', 'bool')).toMatchSnapshot())

    it('should convert object value', () =>
      expect(convertFromGuiValue("{ foo: 'bar' }", 'object')).toMatchSnapshot())

    it('should convert array value', () =>
      expect(convertFromGuiValue('[1, 2]', 'array')).toMatchSnapshot())

    it('should convert symbol value', () =>
      expect(convertFromGuiValue("Symbol('foo')", 'symbol')).toMatchSnapshot())

    it('should convert func value', () =>
      expect(convertFromGuiValue('() => ({})', 'func')).toMatchSnapshot())
  })
})
