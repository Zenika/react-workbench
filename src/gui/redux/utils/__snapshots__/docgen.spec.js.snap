// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gui/redux/utils convert GUI value to model value should convert array value 1`] = `
Array [
  1,
  2,
]
`;

exports[`gui/redux/utils convert GUI value to model value should convert boolean value 1`] = `true`;

exports[`gui/redux/utils convert GUI value to model value should convert enum value 1`] = `"foo"`;

exports[`gui/redux/utils convert GUI value to model value should convert func value 1`] = `[Function]`;

exports[`gui/redux/utils convert GUI value to model value should convert number value 1`] = `0`;

exports[`gui/redux/utils convert GUI value to model value should convert object value 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`gui/redux/utils convert GUI value to model value should convert string value 1`] = `"foo"`;

exports[`gui/redux/utils convert GUI value to model value should convert symbol value 1`] = `"Symbol('foo')"`;

exports[`gui/redux/utils convert docgen prop value to model value should convert array value 1`] = `
Array [
  1,
  2,
]
`;

exports[`gui/redux/utils convert docgen prop value to model value should convert boolean value 1`] = `false`;

exports[`gui/redux/utils convert docgen prop value to model value should convert enum value 1`] = `"foo"`;

exports[`gui/redux/utils convert docgen prop value to model value should convert func value 1`] = `[Function]`;

exports[`gui/redux/utils convert docgen prop value to model value should convert number value 1`] = `0`;

exports[`gui/redux/utils convert docgen prop value to model value should convert object value 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`gui/redux/utils convert docgen prop value to model value should convert string value 1`] = `"foo"`;

exports[`gui/redux/utils convert docgen prop value to model value should convert symbol value 1`] = `"Symbol('foo')"`;

exports[`gui/redux/utils convert docgen to model structure should convert docgen to model 1`] = `
Array [
  Object {
    "name": "name",
    "value": "World",
  },
]
`;

exports[`gui/redux/utils convert model value to GUI value should convert array value 1`] = `"[1,2]"`;

exports[`gui/redux/utils convert model value to GUI value should convert boolean value 1`] = `false`;

exports[`gui/redux/utils convert model value to GUI value should convert enum value 1`] = `"foo"`;

exports[`gui/redux/utils convert model value to GUI value should convert func value 1`] = `[Function]`;

exports[`gui/redux/utils convert model value to GUI value should convert number value 1`] = `0`;

exports[`gui/redux/utils convert model value to GUI value should convert object value 1`] = `"{\\"foo\\":\\"bar\\"}"`;

exports[`gui/redux/utils convert model value to GUI value should convert string value 1`] = `"foo"`;

exports[`gui/redux/utils convert model value to GUI value should convert symbol value 1`] = `Symbol(foo)`;
